fe/fe-core/src/main/java/com/starrocks/lake/StarMgrMetaSyncer.java

```diff
@@ -236,72 +237,143 @@ public int deleteUnusedWorker() {
+            }
         }
+    }
 
-        HashMap<Long, List<Long>> feGroupToShards = new HashMap<>();
+    // return true if starmgr shard meta changed
+    private boolean syncTableMetaInternal(Database db, OlapTable table, boolean forceDeleteData) throws DdlException {
+        StarOSAgent starOSAgent = GlobalStateMgr.getCurrentStarOSAgent();
+        HashMap<Long, Set<Long>> redundantGroupToShards = new HashMap<>();
+        Collection<PhysicalPartition> physicalPartitions = null;
         Locker locker = new Locker();
         locker.lockDatabase(db, LockType.READ);
         try {
-            Table table = db.getTable(tableName);
-            if (table == null) {
-                throw new DdlException(String.format("table %s does not exist.", tableName));
-            }
-            if (!table.isCloudNativeTableOrMaterializedView()) {
-                throw new DdlException("only support cloud table or cloud mv.");
+            if (db.getTable(table.getId()) == null) {
+                return false; // table might be dropped
             }
+            physicalPartitions = table.getAllPhysicalPartitions();
+        } finally {
+            locker.unLockDatabase(db, LockType.READ);
+        }
 
-            OlapTable olapTable = (OlapTable) table;
-            for (Partition partition : olapTable.getAllPartitions()) {
-                for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
-                    long groupId = physicalPartition.getShardGroupId();
-                    List<Long> feShardIds = new ArrayList<>();
-                    for (MaterializedIndex materializedIndex :
-                            physicalPartition.getMaterializedIndices(MaterializedIndex.IndexExtState.ALL)) {
-                        for (Tablet tablet : materializedIndex.getTablets()) {
-                            feShardIds.add(tablet.getId());
-                        }
-                    }
-                    if (!feShardIds.isEmpty()) {
-                        feGroupToShards.put(groupId, feShardIds);
+        for (PhysicalPartition partition : physicalPartitions) {
+            locker.lockDatabase(db, LockType.READ);
+            try {
+                if (table.getState() != OlapTable.OlapTableState.NORMAL) {
+                    return false; // table might be in schema change
+                }
+                PhysicalPartition physicalPartition = table.getPhysicalPartition(partition.getId());
+                if (physicalPartition == null) {
+                    return false; // table changed, wait for next round
+                }
+                long groupId = physicalPartition.getShardGroupId();
+                List<Long> starmgrShardIds = starOSAgent.listShard(groupId);
+                Set<Long> starmgrShardIdsSet = new HashSet<>(starmgrShardIds);
+                for (MaterializedIndex materializedIndex :
+                        physicalPartition.getMaterializedIndices(MaterializedIndex.IndexExtState.ALL)) {
```

Comment:
Maybe we only need to check materialized indexes of state `IndexExtState.VISIBLE`?